pyron-core/src/test/scala/com/cloudentity/pyron/rule/PreparedRewriteTest.scala

```diff
@@ -0,0 +1,164 @@
+      res mustBe List(("stuff", 2), ("ing", 6))
+    }
+    "extract only valid param placeholders" in {
+      // valid param placeholder has non-empty param name surrounded by curly braces, the name may
+      // only contain underscore, small/big latin letters and digits, but can't start with a digit
+      val pattern = "/{ValId_Id}{not an id}/(anon_group_one)/{_1_ok_2_}/(anon_group_two)" ++
+        "{9_starts_with_a_digit_not_an_id}/{}{żółć}{this_ONE_is_valid_ID}"
+      // if chunk within curly braces is not valid param name, it won't be used or interpreted
+      val res = paramNamesWithGroupIndex(pattern, getGroupsCountingPattern(pattern))
+      res mustBe List(("ValId_Id", 1), ("_1_ok_2_", 3), ("this_ONE_is_valid_ID", 5))
+    }
+  }
+
+  "convertNegToPosNumericRefs" should {
+    "replace all negative numeric references with a positive numeric references" in {
+      convertNegToPosNumericRefs("rewrite {-3} has some negative {-1} and positive {2} numeric refs {-2}", 4) mustBe
+        "rewrite {1} has some negative {3} and positive {2} numeric refs {2}"
+    }
+  }
+
+  "insertParamGroupsAndRefs" should {
+    val pattern = "Grand, but (how)/{com}e/I/ca(nn)ot/underst{and}/why/I/have/two/h{and}s/{and}/two/legs!"
+    val rewrite = "/hello/{and}/wel{com}e/to-the-bra{and}-new/gr{and}-coin-toss-{com}petition"
+    val (pat, rew) = insertParamGroupsAndRefs(pattern, rewrite, List(("com", 2), ("and", 5)))
+    "replace first occurrence of param with capture group and remaining occurrences with back-references" in {
+      pat mustBe """Grand, but (how)/([^/]+)e/I/ca(nn)ot/underst([^/]+)/why/I/have/two/h\5s/\5/two/legs!"""
+    }
+    "replace all named param references in rewrite with numeric references" in {
+      rew mustBe """/hello/{5}/wel{2}e/to-the-bra{5}-new/gr{5}-coin-toss-{2}petition"""
+    }
+  }
+
+  "makeParamName" should {
+    "get param name from placeholder string" in {
+      val placeholder = "{some_param_name}"
+      makeParamName(placeholder) mustBe "some_param_name"
+    }
+  }
+
+  "replaceFirstAndRest" should {
+    "replace first occurrence and all remaining occurrences with replacements provided" in {
+      replaceFirstAndRest(
+        haystack = "abcxyz bar boo baz abcxyz goo blah abcxyz",
+        needle = "abcxyz",
+        replaceFirstWith = "ABC",
+        replaceRestWith = "XYZ"
+      ) mustBe "ABC bar boo baz XYZ goo blah XYZ"
+    }
+  }
+
+  "rewritePathWithPathParams" should {
+    "replace param references within path with param values" in {
+      rewritePathWithPathParams("/some/{one}/path/{two}/{one}/", PathParams(Map("one" -> "UNO", "two" -> "DUO"))) mustBe
+        "/some/UNO/path/DUO/UNO/"
+    }
+  }
+
+  "produce rewritten path with resolved params, captured by the matching pattern" in {
+    val preparedRewrite = PreparedRewrite(
+      "/pattern/with_(balanced_(parens))/should/{one}/be/{two}/ok/{two}",
+      "/api",
+      "/resources/{two}/{one}/{1}")
```

Comment:
Ok, for enumarated it's pretty much same like in nginx -> number refers to the number of capture group counting from left side. Named params also count as capture group. For extra convienience, counting from right-side is allowed by using negative number, so {-1} is quite handy for the 'very last capture group' - typically trailing /(.*) .
Nested pattern such as {some {stuff {param} }} is not treated in any special way in that ... only the inner-most {param} can be a param definition, because param name can't contain '{', only [a-zA-Z_0-9]. I'll add info to the docs.